File: js/ice_core.js
// Pure ICE-scoring core (no I/O) for the pmtools `ice` store.
//
// JS twin of py/ice_core.py — mirrors every function 1:1 and is graded against
// the SAME fixtures/ice/*.cases.json the Python harness loads. Where Python
// raises ValueError, this throws Error. Validation-FAILURE fixtures use
// {"name":..., "args":[row], "expected_error": true}.
//
// Scales: I 3/2/1/0.5/0.25 · C 1.0/0.8/0.5 · E 10/7/5/3/1.
// Formula: ICE = I*C*E (higher Ease => higher ICE); tiebreak +1/(issue*1000).
'use strict';

// Empty/None -> null, else String(v). Mirrors py _to_str / lccjs toStr.
function toStr(v) {
  if (v === null || v === undefined || v === '') return null;
  return String(v);
}

// Closed input vocabularies — a value outside the set is rejected.
const VALID_I = [0.25, 0.5, 1, 2, 3];
const VALID_C = [0.5, 0.8, 1.0];
const VALID_E = [1, 3, 5, 7, 10];

// Tier overrides sit above the normal ICE queue (lower = sorts first).
const ICE_TIER_ORDER = { critical: 0, elevated: 1, '': 2 };

// Stored (INSERT) columns — ice_rank is NOT stored; it is derived at export.
const ICE_COLS = [
  'id', 'issue', 'title', 'type', 'I', 'C', 'E', 'ice_score', 'tier',
  'yegor_priority', 'actionable', 'provisional', 'labels', 'notes', 'updated_iso',
];
// CSV export columns — inserts the derived ice_rank after ice_score and drops id,
// matching the lccjs stats/ice-scores.csv this store replaces.
const ICE_CSV_COLS = [
  'issue', 'title', 'type', 'I', 'C', 'E', 'ice_score', 'ice_rank', 'tier',
  'yegor_priority', 'actionable', 'provisional', 'labels', 'notes', 'updated_iso',
];

// ICE = round(I*C*E, 4). Multiplicative (lccjs #1327). Math.round is half-up,
// matching py int(x*10000 + 0.5) for the non-negative ICE domain.
function computeIce(I, C, E) {
  return Math.round(I * C * E * 10000) / 10000;
}

// Tiebreaker: earlier issues win ties but cannot flip a higher score.
function finalScore(ice, issue) {
  return ice + 1 / (issue * 1000);
}

// Stable sort by tier (critical > elevated > normal) then finalScore desc.
function sortRows(rows) {
  return rows.slice().sort((a, b) => {
    const ta = ICE_TIER_ORDER[a.tier || ''] ?? 2;
    const tb = ICE_TIER_ORDER[b.tier || ''] ?? 2;
    if (ta !== tb) return ta - tb;
    return finalScore(b.ice_score || 0, b.issue) - finalScore(a.ice_score || 0, a.issue);
  });
}

// sortRows + a 1-based ice_rank on each row (returns new objects).
function rankRows(rows) {
  return sortRows(rows).map((r, i) => ({ ...r, ice_rank: i + 1 }));
}

// Provisional {I,C,E} from labels alone (the --auto sweep). I from severity,
// C a neutral 0.8, E a coarse guess from the type label. provisional=1.
function deriveAutoScore(labels) {
  const names = (labels || []).filter((l) => l && typeof l === 'object').map((l) => l.name);
  let I;
  if (names.includes('severity:critical')) I = 3;
  else if (names.includes('severity:high')) I = 2;
  else if (names.includes('severity:medium')) I = 1;
  else if (names.includes('severity:low')) I = 0.5;
  else I = 1;
  let E;
  if (names.some((n) => ['documentation', 'chore', 'cleanup', 'style'].includes(n))) E = 7;
  else if (names.some((n) => ['research', 'spike', 'experiment'].includes(n))) E = 3;
  else E = 5;
  return { I, C: 0.8, E };
}

// The override tier from priority:* labels, or '' for the normal queue.
function detectIceTier(labels) {
  const names = (labels || []).filter((l) => l && typeof l === 'object').map((l) => l.name);
  if (names.includes('priority:critical')) return 'critical';
  if (names.includes('priority:elevated')) return 'elevated';
  return '';
}

// null/'' -> null; a value in `valid` -> the number; else throw.
function toIceInput(name, v, valid) {
  if (v === null || v === undefined || v === '') return null;
  const n = Number(v);
  if (!Number.isFinite(n)) throw new Error(`"${name}" must be a number`);
  if (!valid.includes(n)) {
    throw new Error(`"${name}" must be one of ${JSON.stringify(valid)} — got ${v}`);
  }
  return n;
}

// Validate + normalise an ICE row. Required: issue (positive int). I/C/E
// optional but, if given, must be in their valid sets; ice_score computed when
// all three present. Returns the non-id, non-updated_iso ICE_COLS (CLI stamps those).
function validateIceRow(row) {
  if (row === null || typeof row !== 'object' || Array.isArray(row)) {
    throw new Error('ice row must be an object');
  }
  const issue = row.issue;
  if (typeof issue !== 'number' || !Number.isInteger(issue) || issue <= 0) {
    throw new Error('"issue" must be a positive integer');
  }
  const I = toIceInput('I', row.I, VALID_I);
  const C = toIceInput('C', row.C, VALID_C);
  const E = toIceInput('E', row.E, VALID_E);
  const ice_score = (I !== null && C !== null && E !== null) ? computeIce(I, C, E) : null;
  const yp = row.yegor_priority;
  const yegor_priority = Number.isInteger(yp) ? yp : null;

  return {
    issue,
    title: toStr(row.title),
    type: toStr(row.type),
    I,
    C,
    E,
    ice_score,
    tier: toStr(row.tier) || '',
    yegor_priority,
    actionable: toStr(row.actionable) || 'Y',
    provisional: (row.provisional === 1 || row.provisional === true) ? 1 : 0,
    labels: toStr(row.labels),
    notes: toStr(row.notes),
  };
}

module.exports = {
  VALID_I, VALID_C, VALID_E, ICE_TIER_ORDER, ICE_COLS, ICE_CSV_COLS,
  toStr, computeIce, finalScore, sortRows, rankRows,
  deriveAutoScore, detectIceTier, validateIceRow,
};

File: py/ice_core.py
"""Pure ICE-scoring core (no I/O) for the pmtools `ice` store.

Ported faithfully from lccjs scripts/ice-score.js (the standalone scorer this
command replaces). ICE = Impact x Confidence x Ease, a per-issue triage score.

These are the I/O-free seams — the column orders, the score formula, the
ranking, the label-derived --auto heuristic, and the row validator (which RAISES
ValueError on a bad row). The sqlite engine (store.py) and the CLI (ice.py) call
in so the rules are testable without a DB and graded against fixtures/ice/*.

Parity: js/ice_core.js mirrors every function here and is graded against the SAME
fixtures. Validation-FAILURE fixtures use {"name":..., "args":[row],
"expected_error": true} — both languages assert a raised/thrown error.

Scales (verbatim from ice-score.js):
  I (Impact):     3=massive · 2=high · 1=medium · 0.5=low · 0.25=minimal
  C (Confidence): 1.0=high  · 0.8=medium · 0.5=low
  E (Ease):       10=trivial · 7=easy · 5=moderate · 3=hard · 1=very hard
Formula: ICE = I x C x E (higher Ease => higher ICE); tiebreak +1/(issue*1000).
"""


def _to_str(v):
    """Empty/None -> None, else str(v). Mirrors _to_str / lccjs toStr."""
    if v is None or v == "":
        return None
    return str(v)


# Closed input vocabularies — a value outside the set is rejected.
VALID_I = (0.25, 0.5, 1, 2, 3)
VALID_C = (0.5, 0.8, 1.0)
VALID_E = (1, 3, 5, 7, 10)

# Tier overrides sit above the normal ICE queue (lower = sorts first).
ICE_TIER_ORDER = {"critical": 0, "elevated": 1, "": 2}

# Stored (INSERT) columns — ice_rank is NOT stored; it is derived at export.
ICE_COLS = [
    "id", "issue", "title", "type", "I", "C", "E", "ice_score", "tier",
    "yegor_priority", "actionable", "provisional", "labels", "notes", "updated_iso",
]
# CSV export columns — inserts the derived ice_rank after ice_score and drops id,
# matching the lccjs stats/ice-scores.csv this store replaces.
ICE_CSV_COLS = [
    "issue", "title", "type", "I", "C", "E", "ice_score", "ice_rank", "tier",
    "yegor_priority", "actionable", "provisional", "labels", "notes", "updated_iso",
]


def compute_ice(I, C, E):
    """ICE = round(I x C x E, 4). Multiplicative: higher Ease => higher score
    (lccjs #1327 reversed the old `/E` that sank quick wins). Half-up rounding
    to mirror JS Math.round exactly (ICE is always >= 0)."""
    return int(I * C * E * 10000 + 0.5) / 10000


def final_score(ice, issue):
    """Tiebreaker: earlier issues win ties but cannot flip a higher score."""
    return ice + 1 / (issue * 1000)


def sort_rows(rows):
    """Stable sort by tier (critical > elevated > normal) then final_score desc."""
    def key(r):
        tier = ICE_TIER_ORDER.get(r.get("tier") or "", 2)
        return (tier, -final_score(r.get("ice_score") or 0, r["issue"]))
    return sorted(rows, key=key)


def rank_rows(rows):
    """sort_rows + a 1-based `ice_rank` on each row (returns new dicts)."""
    out = []
    for i, r in enumerate(sort_rows(rows)):
        nr = dict(r)
        nr["ice_rank"] = i + 1
        out.append(nr)
    return out


def derive_auto_score(labels):
    """Provisional {I,C,E} from labels alone (the --auto sweep). Rough on
    purpose: I from severity, C a neutral 0.8 (labels can't reveal confidence),
    E a coarse guess from the type label. Rows scored this way are flagged
    provisional=1 for later human review."""
    names = [l.get("name") for l in (labels or []) if isinstance(l, dict)]
    if "severity:critical" in names:
        I = 3
    elif "severity:high" in names:
        I = 2
    elif "severity:medium" in names:
        I = 1
    elif "severity:low" in names:
        I = 0.5
    else:
        I = 1
    if any(n in names for n in ("documentation", "chore", "cleanup", "style")):
        E = 7
    elif any(n in names for n in ("research", "spike", "experiment")):
        E = 3
    else:
        E = 5
    return {"I": I, "C": 0.8, "E": E}


def detect_ice_tier(labels):
    """The override tier from `priority:*` labels, or '' for the normal queue."""
    names = [l.get("name") for l in (labels or []) if isinstance(l, dict)]
    if "priority:critical" in names:
        return "critical"
    if "priority:elevated" in names:
        return "elevated"
    return ""


# The label a tier maps to; 'none' has no label (it clears the overrides).
OVERRIDE_LABEL = {"critical": "priority:critical", "elevated": "priority:elevated"}
VALID_TIERS = ["critical", "elevated", "none"]


def set_tier_plan(tier, current_labels):
    """Pure: plan the label mutation + stored tier for `ice set-tier` (#112),
    given a requested tier and the issue's CURRENT label NAMES (strings). Raises
    ValueError on an invalid tier. critical/elevated -> ensure that priority:*
    label and drop the other override; none -> drop both override labels. On an
    already-correct issue the plan is empty (add None, remove []), so `set-tier
    none` on a clean issue is a no-op. Returns {tier, storedTier, add, remove}:
    storedTier is '' for none (the normal-queue value detect_ice_tier uses), add
    is the label to add or None, remove lists only the override labels present."""
    t = str("" if tier is None else tier).lower()
    if t not in VALID_TIERS:
        raise ValueError("tier must be one of {} — got {}".format(VALID_TIERS, repr(tier)))
    names = [l for l in (current_labels or []) if isinstance(l, str) and l]
    crit = OVERRIDE_LABEL["critical"]
    elev = OVERRIDE_LABEL["elevated"]
    if t == "none":
        return {"tier": t, "storedTier": "", "add": None,
                "remove": [l for l in (crit, elev) if l in names]}
    target = OVERRIDE_LABEL[t]
    other = elev if t == "critical" else crit
    return {"tier": t, "storedTier": t,
            "add": None if target in names else target,
            "remove": [other] if other in names else []}


def _to_ice_input(name, v, valid):
    """None/'' -> None; a value in `valid` -> normalised number; else raise."""
    if v is None or v == "":
        return None
    try:
        n = float(v)
    except (TypeError, ValueError):
        raise ValueError('"{}" must be a number'.format(name))
    if n.is_integer():
        n = int(n)
    if n not in valid:
        raise ValueError(
            '"{}" must be one of {} — got {}'.format(name, list(valid), v))
    return n


def validate_ice_row(row):
    """Validate + normalise an ICE row. Required: issue (positive int). I/C/E
    are optional but, if given, must be in their valid sets; ice_score is
    computed when all three are present. provisional coerces to 0/1. Returns a
    dict keyed by the non-id, non-updated_iso ICE_COLS (the CLI stamps those)."""
    if not isinstance(row, dict):
        raise ValueError("ice row must be an object")

    issue = row.get("issue")
    if isinstance(issue, bool) or not isinstance(issue, int) or issue <= 0:
        raise ValueError('"issue" must be a positive integer')

    I = _to_ice_input("I", row.get("I"), VALID_I)
    C = _to_ice_input("C", row.get("C"), VALID_C)
    E = _to_ice_input("E", row.get("E"), VALID_E)
    ice_score = (compute_ice(I, C, E)
                 if (I is not None and C is not None and E is not None) else None)

    yp = row.get("yegor_priority")
    yegor_priority = yp if (isinstance(yp, int) and not isinstance(yp, bool)) else None

    return {
        "issue": issue,
        "title": _to_str(row.get("title")),
        "type": _to_str(row.get("type")),
        "I": I,
        "C": C,
        "E": E,
        "ice_score": ice_score,
        "tier": _to_str(row.get("tier")) or "",
        "yegor_priority": yegor_priority,
        "actionable": _to_str(row.get("actionable")) or "Y",
        "provisional": 1 if row.get("provisional") in (1, True) else 0,
        "labels": _to_str(row.get("labels")),
        "notes": _to_str(row.get("notes")),
    }
